R/add_cols.R

```diff
@@ -118,60 +118,86 @@ NULL
 }
 
 #' @name .add_date
-.add_deaths <- function(.data,
-                        onset_to_death,
-                        hosp_death_risk,
-                        non_hosp_death_risk) {
+.add_outcome <- function(.data,
+                         onset_to_death,
+                         onset_to_recovery,
+                         hosp_death_risk,
+                         non_hosp_death_risk,
+                         config) {
   infected_idx <- .data$infected == "infected"
   num_infected <- sum(infected_idx)
-  .data$deaths <- NA_real_
-  .data$deaths[infected_idx] <- .data$time[infected_idx] +
-    onset_to_death(num_infected)
+  .data$outcome <- "contact"
+  .data$outcome_time <- NA_real_
+  .data$outcome[infected_idx] <- "recovered"
+  .data$outcome_time[infected_idx] <- .data$time[infected_idx] +
+    onset_to_recovery(num_infected)
+  hosp_idx <- !is.na(.data$hospitalisation) & infected_idx
+  non_hosp_idx <- is.na(.data$hospitalisation) & infected_idx
 
-  apply_death_risk <- function(.data, risk, hosp = TRUE) {
+  # internal function only called in .add_outcome()
+  # assign deaths using population or age-stratified death risk
+  # if risk is NA then no deaths are assigned
+  apply_death_risk <- function(.data, risk, idx, config) {
     if (!is_na(risk)) {
+      # single population risk is a special case of the age-strat risk
+      # convert population risk to data.frame to apply the same operations
       if (is.numeric(risk)) {
-        # size is converted to an integer internally in sample()
-        pop_sample <- sample(
-          which(infected_idx),
-          replace = FALSE,
-          size = (1 - risk) * num_infected
+        risk <- data.frame(
+          min_age = min(.data$age),
+          max_age = max(.data$age),
+          risk = risk
         )
-        .data$deaths[pop_sample] <- NA_real_
-      } else {
-        for (i in seq_len(nrow(risk))) {
-          age_bracket <- risk$min_age[i]:risk$max_age[i]
-          if (hosp) {
-            age_group <- which(
-              .data$age %in% age_bracket & !is.na(.data$hospitalisation)
-            )
-          } else {
-            age_group <- which(
-              .data$age %in% age_bracket & is.na(.data$hospitalisation)
+      }
+      # loop through each age group and sample which cases die
+      for (i in seq_len(nrow(risk))) {
+        age_bracket <- risk$min_age[i]:risk$max_age[i]
+        age_group_idx <- which(.data$age %in% age_bracket & idx)
+
+        if (is.function(config$time_varying_death_risk)) {
+          .check_func_req_args(config$time_varying_death_risk, n_req_args = 2)
+          risk_ <- config$time_varying_death_risk(
+            risk = risk$risk[i],
+            time = .data$time[age_group_idx]
+          )
+          valid_risk <- checkmate::test_numeric(
+            risk_,
+            lower = 0,
+            upper = 1,
+            any.missing = FALSE
+          )
```

Comment:
I wonder whether the calculation of the individual-specific fatality risk and the associated testing on the range of the return could be called fewer times. Here's a reprex to show my thinking - perhaps you're already doing what's suggested here, and it's a tiny performance gain, but I was wondering whether there wasn't some scope for vectorisation. This does depend on `config$time_varying_death_risk()` being vectorised for both `time` and `risk` though.

Then again perhaps my understanding of the intended working here is wrong.

``` r
# arbitrary data
data = withr::with_seed(
  1,
  data.frame(
    name = letters[1:10],
    age = round(runif(10, 0, 100)),
    time_of_infection = seq(0, 95, length.out = 10)
  )
)

# arbitrary baseline fatality risk from data.frame
cfr_table = data.frame(
  age_limit = c(30, 100),
  risk = c(0.01, 0.5) # young have lower risk
)

# assign age bin to data
data$risk = cfr_table$risk[findInterval(data$age, cfr_table$age_limit) + 1]

# get time-dependent risk fn
td_risk = function(time, x) {
  ifelse(time > 50, x / 2, x) # risk reduces at later times
}

# apply time dependence to risk
data$risk = td_risk(data$time_of_infection, data$risk)

# check if values are in expected range
checkmate::assert_numeric(data$risk, lower = 0, upper = 1)

data
#>    name age time_of_infection  risk
#> 1     a  27           0.00000 0.010
#> 2     b  37          10.55556 0.500
#> 3     c  57          21.11111 0.500
#> 4     d  91          31.66667 0.500
#> 5     e  20          42.22222 0.010
#> 6     f  90          52.77778 0.250
#> 7     g  94          63.33333 0.250
#> 8     h  66          73.88889 0.250
#> 9     i  63          84.44444 0.250
#> 10    j   6          95.00000 0.005

# check who actually dies
data$outcome = withr::with_seed(
  1,
  stats::rbinom(n = nrow(data), size = 1, prob = data$risk)
)

data
#>    name age time_of_infection  risk outcome
#> 1     a  27           0.00000 0.010       0
#> 2     b  37          10.55556 0.500       0
#> 3     c  57          21.11111 0.500       1
#> 4     d  91          31.66667 0.500       1
#> 5     e  20          42.22222 0.010       0
#> 6     f  90          52.77778 0.250       1
#> 7     g  94          63.33333 0.250       1
#> 8     h  66          73.88889 0.250       0
#> 9     i  63          84.44444 0.250       0
#> 10    j   6          95.00000 0.005       0
```

<sup>Created on 2024-04-24 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

Reply:
Thanks for this suggestion. It has been implemented in 0295cb2747fb7e6526130757959b85f66d4bd9ea.